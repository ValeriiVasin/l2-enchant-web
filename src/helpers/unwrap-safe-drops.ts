import { EnchantResult } from '../types';
import { calculateEnchant } from './calculate-enchant';

export function unwrapSafeDrops(
  chanceGroups: Array<Array<number>>,
): Array<EnchantResult> {
  const resultGroups = chanceGroups.map((group) =>
    calculateEnchant(group, 'drop'),
  );

  for (let i = 1; i < resultGroups.length; i++) {
    const prevGroup = resultGroups[i - 1];
    const carryEnchants = prevGroup[prevGroup.length - 1].enchants;
    const currentGroup = resultGroups[i];

    for (let j = 0; j < currentGroup.length; j++) {
      currentGroup[j].enchants += carryEnchants;
    }
  }

  return resultGroups.flat();
}
