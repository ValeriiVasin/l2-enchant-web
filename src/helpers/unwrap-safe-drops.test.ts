import { test, expect } from '@jest/globals';
import { unwrapSafeDrops } from './unwrap-safe-drops';

test('unwraps safe drops', () => {
  expect(
    unwrapSafeDrops([
      [50, 50],
      [25, 25],
    ]),
  ).toMatchSnapshot();
});
