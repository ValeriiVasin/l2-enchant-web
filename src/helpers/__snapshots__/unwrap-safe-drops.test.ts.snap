// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unwraps safe drops 1`] = `
[
  {
    "chance": 50,
    "enchants": 2,
    "items": 1,
  },
  {
    "chance": 50,
    "enchants": 6,
    "items": 1,
  },
  {
    "chance": 25,
    "enchants": 10,
    "items": 1,
  },
  {
    "chance": 25,
    "enchants": 26,
    "items": 1,
  },
]
`;
